### PR TITLE
Fix various windows build issues

### PR DESCRIPTION
**Pull request by Keno**
@trilinos/epetra @trilinos/teuchos @trilinos/ifpack

## Motivation

Fixes a number of minor issues when building for windows (in particular a MinGW cross compile from linux - though most of the issues are applicable to MSVC also). Details of each particular change are in the commit message. With these changes, my specific configuration of Trilinos builds fully for Windows.

## Testing

Automated build testing for all platforms I have access to was done in https://github.com/JuliaPackaging/Yggdrasil/pull/3569. No functional changes expected.

### CODE REVIEW

**Comment by jhux2** [comment]
@trilinos/epetra @trilinos/teuchos @trilinos/ifpack

**Comment by jhux2** [comment]
I haven't reviewed, but I will approve to let the AT run.  **DO NOT MERGE** until reviewed.

**Review by jhux2 — APPROVED** [review]
Approving to let the autotester run.

File: cmake/tribits/core/package_arch/TribitsConstants.cmake
Comment (bartlettroscoe) [inline cmake/tribits/core/package_arch/TribitsConstants.cmake:99]: What happens on Cygwin?

Comment (Keno) [reply cmake/tribits/core/package_arch/TribitsConstants.cmake:99]: CMAKE_HOST_SYSTEM_NAME is CYGWIN on cygwin. Up to date versions of CMake do not define WIN32 on cygwin unless cross compiling using mingw (which is the correct behavior, because cygwin proper is POSIX not WIN32). As a result this change should not affect behavior on cygwin (and in particular, if cygwin git is installed it should work here). The one caveat there is that older versions of CMake do incorrectly define WIN32 on cygwin. I believe the behavior as implemented is correct, because building on cygwin, you do want to use the cygwin git build if at all possible to pick up any cygwin specific representations (e.g. symbolic links). It is possible that invoking the WIN32 git would work also, but it's be a bit of a Frankenstein thing to do.

File: packages/teuchos/parameterlist/src/Teuchos_ParameterListAcceptorDefaultBase.hpp
Comment (bartlettroscoe) [inline packages/teuchos/parameterlist/src/Teuchos_ParameterListAcceptorDefaultBase.hpp:61]: Okay, we are just adding an export.  I can't see how this can break an existing windows build.

File: cmake/tribits/core/package_arch/TribitsConstants.cmake
Comment (bartlettroscoe) [inline cmake/tribits/core/package_arch/TribitsConstants.cmake:106]: Okay, we are just fixing the logic for modern CMake versions on Windows and this will still use 'git' for cygwin.

File: packages/teuchos/numerics/src/Teuchos_BLAS.cpp
Comment (bartlettroscoe) [inline packages/teuchos/numerics/src/Teuchos_BLAS.cpp:120]: This is inside of an ifdef for `_MSC_VER` so Trilinos PR testing will **not** test these changes.  But this looks like correct C++ code (and I am surprised that it compiled the way it was).

File: packages/epetra/src/Epetra_C_wrappers.h
Comment (bartlettroscoe) [inline packages/epetra/src/Epetra_C_wrappers.h:59]: This seems reasonable.  C++11 added official support for `long long` int and it is guaranteed to be at least 64 bits.  So I would think this is the right change to make to achieve that.

File: packages/ifpack/src/Ifpack_SerialTriDiSolver.h
Comment (bartlettroscoe) [inline packages/ifpack/src/Ifpack_SerialTriDiSolver.h:130]: Will removing this export not break some customer's usage of this class from the DLL?  Why is this export being removed?  Why not leave this export?

**Review by bartlettroscoe — DISMISSED** [review]
I am good with the TriBITS and Teuchos changes (which is what I am approving here).   I personally think that the change to `Epetra_C_wrappers.h` is okay as well that changes to `long long` for 64 bit integers (but will let some @trilinos/epetra person approve that).

All of the rest of the changes all appear to be just changes to DLL exports.  All of the changes are adding exports except for one case where an export is removed (for the class `Ifpack_SerialTriDiSolver` in the file `Ifpack_SerialTriDiSolver.h`).  Except for that one export removal,  I can't see how adding additional exports can break anything.

Others will need to review and approve the remaining changes for Epetra and Ifpack but @trilinos/ifpack developers should look carefully at the export being removed in the file `Ifpack_SerialTriDiSolver.h` and make sure that is okay.

NOTE: After this PR is merged, I will cherry-pick off the TriBITS change and apply it back to the TriBITS repo and snasphot TriBITS back to Trilinos to address any possible future merge conflicts.

File: packages/ifpack/src/Ifpack_SerialTriDiSolver.h
Comment (Keno) [reply packages/ifpack/src/Ifpack_SerialTriDiSolver.h:130]: As described in the commit message, the export indicates epetra, but this class is in ifpack - as a result this was never exported in the first place and the only effect it has is breaking the build of ifpack on windows. It is certainly possible that this should have an ifpack export, but I felt this to be the more conservative change.

Comment (bartlettroscoe) [reply packages/ifpack/src/Ifpack_SerialTriDiSolver.h:130]: Okay, I see that in commit c4d2092687991bfc657b0440ab78d2652a39bec4.  I should have looked at that.

With that, exports are only being added in this PR so I don't see any reason not to merge this PR.

**Comment by bartlettroscoe** [comment]
@Keno, if no-one else approves this in the next few days, ping back here with a comment and CC those of us that commented here already.

**Comment by bartlettroscoe** [comment]
@trilinos/framework, are there any automated builds of Trilinos for Windows that might be impacted (for better or worse) by this PR?

**Comment by jhux2** [comment]
@jrobcary Does this PR affect your work?

**Comment by jrobcary** [comment]
It looks innocuous, so no objection here.  We have stayed with what we got working.

**Comment by bartlettroscoe** [comment]
@trilinos/framework, @trilinos/epetra, @trilinos/ifpack, @trilinos/ifpack, there is now a conflict here (likely with PR #9663 that was just merged).  Unless there are objections, I propose we resolve this conflict, run another round of PR testing and merge.

**Comment by jhux2** [comment]
> there is now a conflict here (likely with PR #9663 that was just merged). Unless there are objections, I propose we resolve this conflict, run another round of PR testing and merge.

@bartlettroscoe I'm in favor of this.

**Review by kddevin — APPROVED** [review]
The Epetra changes look harmless.  I don't know what EPETRA_LIB_DLL_EXPORT does, and it certainly clutters the code, but the changes here match existing use of EPETRA_LIB_DLL_EXPORT in functions and classes in Epetra.

**Comment by jhux2** [comment]
@Keno If you are able to resolve the conflicts, then this will be ready to merge.

**Comment by Keno** [comment]
> @Keno If you are able to resolve the conflicts, then this will be ready to merge.

Done

**Comment by Keno** [comment]
> I don't know what EPETRA_LIB_DLL_EXPORT does

It marks DLL visibility for windows. Classes without this attribute will not be visible from downstream dlls. Other platforms have a similar concept, but with the opposite default.

**Review by jhux2 — APPROVED** [review]
lgtm

**Comment by bartlettroscoe** [comment]
FYI: I pulled off the commit patch from commit 5dee9af974810141b7f1ea54a21ee3e6998f380e and applied to the TriBITS repo in https://github.com/TriBITSPub/TriBITS/pull/422 (along with fixing the command case).  This will create a merge conflict on the next TriBITS snapshot back to Trilinos but it will be trivial to resolve.

**Pull request merged**
